### PR TITLE
add usdrate service

### PR DESCRIPTION
**Pull request by secmask**
this MR implement USDRate service
- rate of today will be use coingecko, in case of fail, fallback to coinlib
- rate before today will be fetch from DB, if not found will be fallback to coingecko

because coinlib has rate limit, it will keep a cache of rate in 5 min.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
-  directories:
-  - ${TRAVIS_HOME}/gometalinter/gometalinter-${GOMETALINTER_VERSION}-linux-amd64
+before_script:
+  - psql -c "CREATE DATABASE tokenrate;" -U postgres
+  - psql -c "CREATE USER tokenrate WITH PASSWORD 'tokenrate';" -U postgres
+  - psql -c "ALTER ROLE tokenrate SUPERUSER;" -U postgres
 
 script:
-  - gometalinter --config=gometalinter.json ./...
-  - go build -v -mod=vendor ./...
```
Comment (ducquangkstn) [inline .travis.yml:18]: we shouldn't ignore failed to compile

File: usdrate/cmd/usdrate-api/main.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+			Name:   bindAddressFlag,
+			Usage:  "Address to serve ETH price endpoint",
+			Value:  defaultBindAddress,
+			EnvVar: "BIND_ADDRESS",
+		},
+	)
+
+	a.Flags = append(a.Flags, app.NewPostgreSQLFlags("tokenrate")...)
+	a.Flags = append(a.Flags, app.NewSentryFlags()...)
+	a.Flags = append(a.Flags, coinlib.NewFlags()...)
```
Comment (ducquangkstn) [inline usdrate/cmd/usdrate-api/main.go:41]: we need to add flags of CoinGecko

File: usdrate/cmd/usdrate-crawler/main.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		log.Fatal(err)
+	}
+}
+
+func validateTime(fromTimeS, toTimeS string) (time.Time, time.Time, error) {
+	var (
+		fromTime, toTime time.Time
+		err              error
+		now              = time.Now()
+		currentTime      = time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
```
Comment (ducquangkstn) [inline usdrate/cmd/usdrate-crawler/main.go]: now.Truncate() is better since it is UTC

File: usdrate/cmd/usdrate-crawler/main.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+		return crawlTokenPriceWithTimeRange(sugar, fromTime, toTime, ps, s)
+	}
+	logger.Info("to-time is blank, get history price from from-time and run get price daily...")
+	if err := crawlTokenPriceWithTimeRange(sugar, fromTime, toTime, ps, s); err != nil {
+		logger.Errorw("failed to get rate with time range", "from-time", fromTime, "to-time", toTime)
+		return err
+	}
+	if err := crawlTokenPriceDaily(sugar, ps, s, c.String(jobRunningTimeFlag)); err != nil {
+		logger.Errorw("failed to get rate daily", "error", err)
+	}
```
Comment (ducquangkstn) [inline usdrate/cmd/usdrate-crawler/main.go]:
```suggestion
	if err := crawlTokenPriceWithTimeRange(sugar, fromTime, toTime, ps, s); err != nil {
		logger.Errorw("failed to get rate with time range", "from-time", fromTime, "to-time", toTime)
		return err
	}
	if len(toTimeS) != 0 {
	        return nil
	}
        logger.Info("to-time is blank, get history price from from-time and run get price daily...")
        if err := crawlTokenPriceDaily(sugar, ps, s, c.String(jobRunningTimeFlag)); err != nil {
		logger.Errorw("failed to get rate daily", "error", err)
	}
```

File: usdrate/cmd/usdrate-crawler/main.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+	cs := make(chan os.Signal, 1)
+	signal.Notify(cs, os.Interrupt)
+	<-cs
```
Comment (ducquangkstn) [inline usdrate/cmd/usdrate-crawler/main.go:172]: should we add an error channel and also terminate if crawl price daily got errror

File: usdrate/storage/postgres/postgres_test.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
+package postgres
+
+import (
+	"testing"
+	"time"
+
+	_ "github.com/lib/pq" // sql driver name: "postgres"
```
Comment (ducquangkstn) [inline usdrate/storage/postgres/postgres_test.go]: I don't think we need to import this cuz we have it in `testutil.MustNewDevelopmentDB()`

**Review by ducquangkstn — CHANGES_REQUESTED** [review]
Just need some minor changes

File: usdrate/cmd/usdrate-api/main.go (hunk 2 above)
Comment (secmask) [reply usdrate/cmd/usdrate-api/main.go:41]: coingecko currently have no preference that need flags (on other side coinlib need API Key)

File: .travis.yml (hunk 1 above)
Comment (secmask) [reply .travis.yml:18]: I will it backed, removed by a miss take

File: coingecko/flags.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+const (
+	reqTimeWaitingFlag    = "coingecko-req-waiting-time"
+	defaultReqTimeWaiting = time.Second
+)
+
+// NewFlags return cli config for coingecko
+func NewFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.DurationFlag{
```
Comment (NNagato) [inline coingecko/flags.go]: this flag no longer uses in code base, could we remove it?

File: usdrate/cmd/usdrate-api/main.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		cli.StringFlag{
+			Name:   bindAddressFlag,
+			Usage:  "Address to serve ETH price endpoint",
+			Value:  defaultBindAddress,
+			EnvVar: "BIND_ADDRESS",
+		},
+	)
+
+	a.Flags = append(a.Flags, app.NewPostgreSQLFlags("tokenrate")...)
+	a.Flags = append(a.Flags, app.NewSentryFlags()...)
```
Comment (NNagato) [inline usdrate/cmd/usdrate-api/main.go]: sentry flags already added in `NewAppWithMode` so we can remove it here

File: usdrate/cmd/usdrate-crawler/main.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+			Name:   providerFlag,
+			Usage:  "provide provider to get price [coingecko]",
+			EnvVar: "PROVIDER",
+		},
+	)
+	defaultPGDB := "tokenrate"
+	a.Flags = append(a.Flags, coingecko.NewFlags()...)
+	a.Flags = append(a.Flags, app.NewPostgreSQLFlags(defaultPGDB)...)
+	a.Flags = append(a.Flags, app.NewSentryFlags()...)
+	if err := a.Run(os.Args); err != nil {
```
Comment (NNagato) [inline usdrate/cmd/usdrate-crawler/main.go:66]: sentry flags already added in `NewAppWithMode` so we can remove it here

Comment (secmask) [reply usdrate/cmd/usdrate-crawler/main.go:66]: I don't see it in NewAppWithMode?

Comment (NNagato) [reply usdrate/cmd/usdrate-crawler/main.go:66]: My mistake with `NewApp` function, conversation resolved

**Review by NNagato — APPROVED** [review]

**Pull request merged**